### PR TITLE
Fix hanging during `carthage update`

### LINKED ISSUE #445: Issues with 0.7.1 version
I am having multiple issues since I have updated to latest version (0.7.1).
First I have the same issue as #444 but even if I delete Realm from my Cartfile I am still having strange behavior as Fetching or Checking out that never ends, or permissions denied errors.

My Cartfile is very simple:

    github "Alamofire/Alamofire"
    github "Hearst-DD/ObjectMapper"

Here is an example of permission error I had:

capture d ecran 2015-05-08 a 13 15 27

After deleting Carthage folder and Cartfile.resolved file, I am stuck on ```Fetching``` step, but before it was the case for the ```Checking out``` step:

capture d ecran 2015-05-08 a 13 24 11

Is there any known issues about this? 
What can I do to give you more informations to solve this problem?

Since this morning I am not able to retrieve dependencies for my new project.

### LINKED ISSUE #455: Deadlock when fetching reddift by `carthage update --use-submodules`
Reproducing log as following:
```sh
norio-nomura-MacBook-Pro% carthage version
0.7.3
norio-nomura-MacBook-Pro% git --version
git version 2.3.2 (Apple Git-55)
norio-nomura-MacBook-Pro% git clone -b refactoring https://github.com/sonsongithub/reddift.git
Cloning into 'reddift'...
remote: Counting objects: 1402, done.
remote: Compressing objects: 100% (148/148), done.
remote: Total 1402 (delta 71), reused 0 (delta 0), pack-reused 1252
Receiving objects: 100% (1402/1402), 499.31 KiB | 312.00 KiB/s, done.
Resolving deltas: 100% (882/882), done.
Checking connectivity... done.
norio-nomura-MacBook-Pro% cd reddift
norio-nomura-MacBook-Pro% carthage update --use-submodules
*** Fetching Nimble
*** Fetching Quick
*** Checking out Quick at "v0.3.1"
*** Checking out Nimble at "v0.4.2"
*** xcodebuild output can be found in /var/folders/kt/2mwy9b_56_7993x190pl_1fh0000gn/T/carthage-xcodebuild.VakYEe.log
*** Building scheme "Nimble-iOS" in Nimble.xcodeproj
*** Building scheme "Nimble-OSX" in Nimble.xcodeproj
*** Building scheme "Quick-iOS" in Quick.xcworkspace
*** Building scheme "Quick-OSX" in Quick.xcworkspace
norio-nomura-MacBook-Pro% carthage update --use-submodules
*** Fetching Nimble
*** Fetching Quick
^C
norio-nomura-MacBook-Pro% 
```
After killing `carthage` by hitting `^C`, `carthage checkout` and `carthage build` works.

### PR DESCRIPTION
**Pull request by jspahrsummers**
_Depends on https://github.com/Carthage/ReactiveTask/pull/23._

:pray: Fixes #455. Fixes #445. :pray:

**To do:**

 - [ ] Merge dependencies and bump Cartfile
 - [x] Fix fetching after checkout
 - [x] Fix nondeterministic `NoTaggedVersion` and `RequiredVersionNotFound` errors
 - [x] Fix crashes

### CODE REVIEW

File: Cartfile [hunk 1]
```diff
@@ -1,2 +1,2 @@
-github "Carthage/ReactiveTask" ~> 0.5.1
+github "Carthage/ReactiveTask" "upgrade-rac"
```
Comment (robrix) [inline Cartfile]: This has been merged now.

**Comment by bassrock** [comment]
:+1:

**Comment by jspahrsummers** [comment]
Now I have an honest-to-god deadlock, in part because of https://github.com/Carthage/Carthage/pull/472, which allows queues to block each other. :unamused:

**Comment by jspahrsummers** [comment]
https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2032 and https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2029 together fix the crashes here.

**Comment by schwa** [comment]
Any chance this is going to be resolved soon? Want to move current (work) project to Carthage but can't really convince the team not to worry about the hangs…

**Comment by jspahrsummers** [comment]
@schwa We're working on it. In the meantime, [0.6.5](https://github.com/Carthage/Carthage/releases/tag/0.6.5) (or 0.6.4, if you want to use Homebrew) don't exhibit the hangs, so you may want to use those for now.

**Comment by jspahrsummers** [comment]
@Carthage/carthage This is ready for review (though the `Cartfile` has yet to be updated). Note the dependency on https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2042.

**Comment by jspahrsummers** [comment]
I've tagged [0.7.4-beta.1](https://github.com/Carthage/Carthage/releases/tag/0.7.4-beta.1) from this branch, if anyone wants to try it out before it's merged and officially released.

File: Source/CarthageKit/ProducerQueue.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+//  Carthage
+//
+//  Created by Justin Spahr-Summers on 2015-05-23.
+//  Copyright (c) 2015 Carthage. All rights reserved.
+//
+
+import ReactiveCocoa
+
+/// Serializes the execution of SignalProducers, like flatten(.Concat), but
+/// without all needing to be enqueued in the same context.
```
Comment (robrix) [inline Source/CarthageKit/ProducerQueue.swift]: Could you describe why this is important in the comment, please?

**Comment by bassrock** [comment]
Can confirm that the tag worked, but at the end throws a segmentation fault 11, but everything seems to be there and no build errors

**Comment by bassrock** [comment]
Also getting a segmentation fault 11 that is failing the at the copy-frameworks run script.

**Comment by robrix** [comment]
infomercial woman plunging a toilet until she is suddenly splashed with horrible backspray

File: Source/CarthageKit/ProducerQueue.swift (hunk 2 above)
Comment (jspahrsummers) [inline Source/CarthageKit/ProducerQueue.swift]: Addressed in 2d1addc.

File: Source/carthage/swift-is-crashy.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	"Unfortunately, this is probably a bug in Swift and not Carthage.\n"
+	"If this is preventing you from doing work, please file an issue and\n"
+	"we'll do our best to work around it:\n"
+	"\033[4mhttps://github.com/Carthage/Carthage/issues/new\033[0m\n\n";
+
+static void uncaughtSignal(int zig, siginfo_t *info, void *context) {
+	const char *signalName = strsignal(zig);
+	fprintf(stderr, exceptionMessage, signalName);
+
+	raise(zig); // for great justice
```
Comment (jspahrsummers) [inline Source/carthage/swift-is-crashy.c]: cc @keithduncan

**Comment by jspahrsummers** [comment]
@bassrock @norio-nomura Unfortunately, there's not much we can do about those segfaults. Swift is crashy. :disappointed: 

I've added some better logging around it, at least.

File: Source/carthage/swift-is-crashy.c (hunk 3 above)
Comment (keithduncan) [inline Source/carthage/swift-is-crashy.c]: errormessage5

**Comment by norio-nomura** [comment]
Is waiting termination of `grep` before calling `exit()` needed?

**Comment by jspahrsummers** [comment]
@norio-nomura I think build errors could get truncated if we don't wait.

**Comment by norio-nomura** [comment]
@jspahrsummers semaphore of `single()` that main thread is waiting is triggered from `terminationHandler` of last xcodebuild task. That thread continues disposing while main thread calls `exit()` concurrently.
I pushed some code for fixing that scenario.
On my testing, segmentation fault on termination go away.

**Comment by jspahrsummers** [comment]
Thanks to significant help from @norio-nomura, the segfaults should now be fixed (after the refactoring described in https://github.com/Carthage/ReactiveTask/pull/23).

I've released [0.7.4-beta.2](https://github.com/Carthage/Carthage/releases/tag/0.7.4-beta.2) with these proposed fixes as well, if anyone wants to test it out.

**Comment by norio-nomura** [comment]
@jspahrsummers  I'm glad to helping you :smile: 
By the way, bffac3c is tagged on master branch

**Comment by jspahrsummers** [comment]
@norio-nomura Ah, thanks. I've fixed it.

**Comment by jspahrsummers** [comment]
Not sure if it's new with this PR, but I saw evidence of Carthage building dependencies out of order (and thereby failing to build dependents) on 0.7.4b2. I'll investigate.

File: Source/carthage/Build.swift [hunk 4]
```diff
@@ -20,74 +20,91 @@ public struct BuildCommand: CommandType {
...
+								break
+							}
+						}, terminated: {
+							sendCompleted(stdoutSink)
+						}, interrupted: {
+							sendInterrupted(stdoutSink)
+						})
+
+					buildProgress = SignalProducer(values: [ grepTask, buildProgress ])
+						|> flatten(.Merge)
```
Comment (norio-nomura) [inline Source/carthage/Build.swift]: Does `grepTask` termination be waited at somewhere?
Test with https://github.com/Carthage/ReactiveTask/commit/18cdc3fc44e65a1564bc3ff201d49230eae7a986 cause segmentation fault and seems `grepTask`'s threads are working.

**Comment by jspahrsummers** [comment]
There are still issues here, but I think it'd be good to make a _less_ broken Carthage release available. I'd like to merge this, and continue to investigate the remaining issues separately.

After this is merged, you can get the latest version (if you don't want to build it yourself) with `brew install --HEAD carthage`. You can also install [0.6.5](https://github.com/Carthage/Carthage/releases/tag/0.6.5) if you want the last stable version before we started seeing these problems.

**Comment by mdiep** [comment]
> and continue to investigate the remaining issues separately—with which help would be greatly appreciated

Are there specific Issues that capture the remaining problems? I'm a little confused about which things have and haven't been fixed.

**Comment by jspahrsummers** [comment]
@mdiep I guess https://github.com/Carthage/Carthage/issues/455#issuecomment-106991205 and https://github.com/Carthage/Carthage/pull/474/files#r31377774 capture the remaining issues? @norio-nomura might know more.

It's hard to say, because all of the segfaults are bugs in the Swift toolchain. As far as I know, I've worked around all of the ones that I've been able to find, but they're not strictly deterministic, and I don't think there's a one-size-fits-all solution.

Ultimately, I'm pretty burnt out on this, because it's like playing crash whack-a-mole with a broken hammer (which can be Swift and/or LLDB in this analogy).

**Comment by mdiep** [comment]
Maybe it's worth waiting a week to see what WWDC brings?

**Comment by jspahrsummers** [comment]
From what I've seen so far, major updates to the Swift toolchain tend to introduce more issues than they fix. (The fixes usually arriving in patch updates.)

**Comment by norio-nomura** [comment]
I experienced two problems after https://github.com/Carthage/ReactiveTask/commit/18cdc3fc44e65a1564bc3ff201d49230eae7a986.

## Segmentation fault

It seems my probing codes triggered potential problem of https://github.com/Carthage/Carthage/pull/474/files#r31377774. With no probing code, I have not experienced segmentation fault without it.

But I have confirmed that `grepTask` is not waited following steps.
On using `Cartfile`:
  ```
  github "specta/specta" ~> 1.0.0
  ```
  1. Set breakpoint at Extensions.swift:111
  2. Debug run carthage scheme with working directory that contains above `Cartfile`
  3. After hitting breakpoint of 1, see `/usr/bin/grep --extended-egrep (warning|error|failed): serial` queue is still running codes as following screenshot
screenshot 2015-06-01 13 39 57

I don't know another condition such as some `Cartfile` configuration could cause this segmentation fault or not.

## Carthage hangs https://github.com/Carthage/Carthage/issues/455#issuecomment-106991205

Without my probing code, this issue frequently happen. it happens after `*** Fetching` or `*** Checking out`
e.g.
```sh
norio-nomura-MacBook-Pro% /Users/norio/Library/Developer/Xcode/DerivedData/Carthage-bliqitahmqdopabqxfgvfxzydaql/Build/Products/Release/carthage update
*** Fetching Nimble
*** Fetching Quick
*** Downloading Quick at "v0.3.1"
*** Checking out Nimble at "v0.4.2"
```
on stopping above, no child process is working:
screenshot 2015-06-01 13 58 28
Attach debugger to it
screenshot 2015-06-01 14 03 09
no working thread is alive.

These still happen with 9268fe3
Should I file separated issues?

**Comment by norio-nomura** [comment]
I just push PR for fixing the Carthage hangs that I reported above comment and https://github.com/Carthage/Carthage/issues/455#issuecomment-106991205

The PR does not fix the `grepTask` is not waited on termination.
On testing the PR, I got segmentation fault with `Cartfile` of #497 as following:
```sh
norio-nomura-MacBook-Pro% /Users/norio/Library/Developer/Xcode/DerivedData/Carthage-bliqitahmqdopabqxfgvfxzydaql/Build/Products/Release/carthage update
*** Fetching expecta
*** Fetching specta
*** Fetching WLXBluetoothDevice
*** Fetching ReactiveCocoa
*** Fetching Result
*** Fetching Box
*** Checking out WLXBluetoothDevice at "0b834561e4e6551a76c2fec1ae389a404b93d6b9"
*** Checking out Result at "0.4.3"
*** Checking out specta at "v1.0.1"
*** Checking out ReactiveCocoa at "b15798ae4f6386b8461b3b06f49908f2009a28ce"
*** Checking out expecta at "v1.0.1"
*** Downloading Box at "1.2.2"
*** xcodebuild output can be found in /var/folders/kt/2mwy9b_56_7993x190pl_1fh0000gn/T/carthage-xcodebuild.dFcsYM.log
*** Building scheme "Expecta" in Expecta.xcodeproj
*** Building scheme "Expecta-iOS" in Expecta.xcodeproj
*** Building scheme "Specta" in Specta.xcworkspace
*** Building scheme "Specta-iOS" in Specta.xcworkspace
*** Building scheme "Result-iOS" in Result.xcworkspace
*** Building scheme "Result-Mac" in Result.xcworkspace
*** Building scheme "ReactiveCocoa iOS" in ReactiveCocoa.xcworkspace
*** Building scheme "ReactiveCocoa Mac" in ReactiveCocoa.xcworkspace
*** Building scheme "WLXBluetoothDevice" in WLXBluetoothDevice.xcodeproj
** BUILD FAILED **


The following build commands failed:
	Check dependencies
(1 failure)
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	Check dependencies
(1 failure)

ld: warning: embedded dylibs/frameworks only run on iOS 8 or later
ld: warning: embedded dylibs/frameworks only run on iOS 8 or later
ld: warning: embedded dylibs/frameworks only run on iOS 8 or later
ld: warning: embedded dylibs/frameworks only run on iOS 8 or later
Code Sign error: No code signing identities found: No valid signing identities (i.e. certificate and private key pair) matching the team ID “7XS5GQGS29” were found.
CodeSign error: code signing is required for product type 'Framework' in SDK 'iOS 8.3'

Caught signal triggered by the Swift runtime!
Segmentation fault: 11

Unfortunately, this is probably a bug in Swift and not Carthage. If
this is preventing you from doing work, please file an issue and we'll
do our best to work around it:
https://github.com/Carthage/Carthage/issues/new

Please also consider filing a radar with Apple, containing the version
of Carthage and any crash report found in Console.app.

[1]    9084 segmentation fault   update
norio-nomura-MacBook-Pro% 
```

I believe it needs change to `grepTask` being waited on termination for fixing the segmentation fault.

**Comment by norio-nomura** [comment]
Ah, segmentation fault of Carthage seems happen after task completed event and before task disposed, and not only on `grepTask` but also on xcodebuild task.
Waiting completion as same as that https://github.com/Carthage/ReactiveTask/pull/21 did, it should be waiting disposal.
I don't know RAC can wait disposal.
If it can't, https://github.com/Carthage/ReactiveTask/pull/21 will be needed.

**Comment by norio-nomura** [comment]
@jspahrsummers @mdiep As I tested, my concerned issuess since I reported #455 are fixed by following PRs.
hangs:
https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2058
segmentation fault:
https://github.com/Carthage/ReactiveTask/pull/24
https://github.com/Carthage/Carthage/pull/519

I just uploaded private build to https://github.com/norio-nomura/Carthage/releases/tag/0.7.4-privatebuild.1 including those PRs.

**Comment by norio-nomura** [comment]
I just replaced private build package with Developer ID signed version.

**Comment by mdiep** [comment]
@jspahrsummers is better qualified to review those PRs :confused:

**Comment by robrix** [comment]
Since @jspahrsummers is pretty busy at the moment, I’ll take a shot at reviewing #519.

**Comment by jspahrsummers** [comment]
@norio-nomura Thank you so much for all the fixes—they work great! :sparkling_heart: 

I've tagged [0.7.4](https://github.com/Carthage/Carthage/releases/tag/0.7.4), which should fix all the hanging and crashing issues that we know about. The update should [become available in Homebrew](https://github.com/Homebrew/homebrew/pull/40450) soon.

**Comment by norio-nomura** [comment]
You're welcome.
Now, I can recommend Carthage to my friends again. :smile:

**Pull request merged**